### ISSUE TITLE
Storybook works different from production build

### ISSUE THREAD

**Issue by maxatdetroit** [CityOfDetroit/COD-Design-System#134]
**Describe the bug**

When using `<cod-table>` with `data-stacked="true"` attribute set, the design system component works different on storybook vs. production package. Specifically, the component does not properly pass the `table-stacked` class down the element stack on production.

**To Reproduce**

1. Check branch `demo.issue.134`
2. Open `src/html/demo-page-defer.html` in browser
3. Open http://localhost:6006/?path=/story/components-organisms-table--stacked-inline in browser
5. Resize both to mobile width
6. Observe they behave differently
7. Also observe the lifecycle events logged are in different order / frequency

https://github.com/CityOfDetroit/COD-Design-System/assets/143553259/e01cc843-999e-47a8-bce5-9673a0d3a8db

[storybook-console-export-2023-12-14_18-32-48.txt](https://github.com/CityOfDetroit/COD-Design-System/files/13679263/storybook-console-export-2023-12-14_18-32-48.txt)

https://github.com/CityOfDetroit/COD-Design-System/assets/143553259/3a976737-7e54-47f2-8517-576d3f36d8d5

[demo-console-export-2023-12-14_18-32-48.txt](https://github.com/CityOfDetroit/COD-Design-System/files/13679261/demo-console-export-2023-12-14_18-32-48.txt)

[detroitmi-console-export-2023-12-14_18-32-48.txt](https://github.com/CityOfDetroit/COD-Design-System/files/13679262/detroitmi-console-export-2023-12-14_18-32-48.txt)

**Expected behavior**

The table component should work similarly to the styling in storybook. Additionally, lifecycle events should be triggered in the same order / frequency between storybook and production.

**Comment by maxatdetroit** [comment]
### Root Cause

The custom elements `cod-table`, `cod-table-body`, and `cod-table-row` are added to the DOM in the following order:
1. `cod-table`
2. `cod-table-body`
3. `cod-table-row`

Those same elements are defined in `index.js` in the following order:
1. `cod-table-body`
2. `cod-table-row`
3. `cod-table`

This difference in ordering is inconsequential when the script `index.js` is loaded before the DOM is parsed because the order of events is as follows:
1. Load the script: define all custom elements (but don't execute constructors, connectedCallbacks, etc. because elements are not yet declared in the DOM).
2. Parse the DOM: parse tag for `cod-table`, `cod-table-body`, `cod-table-row` (in that order) and call the constructor and `connectedCallback` at each tag parsed. Thus the order of execution for constructors and `connectedCallback`s is `cod-table`, `cod-table-body`, `cod-table-row` and all is well.

However, this difference in DOM parsing order and element definition is consequential when the script is deferred because then the order of the events is as follows:
1. Parse the DOM: parse tag for `cod-table`, `cod-table-body`, `cod-table-row` (in that order). However, the constructor and `connectedCallback` at each tag are not called because they are not yet defined. The browser queues an event to call the element constructors and `connectedCallback`s when they are [defined (i.e. upgraded) later](https://web.dev/articles/custom-elements-v1#element_upgrades).
2. Load the script: define custom elements in the order defined in the script which is `cod-table-body`, `cod-table-row`, then `cod-table` and execute constructors and `connectedCallbacks` at each definition / upgrade (since they are already in the DOM). Thus the order of execution for constructors and `connectedCallback`s is `cod-table-body`, `cod-table-row`, then `cod-table` and the `table-stacked` class fails to propagate all the way to the table row.

### Possible Solutions
1. Change order of definition in the design system source code (e.g. explicitly define main import file with ordered custom element definitions).
2. Implement [reactive properties](https://lit.dev/docs/components/properties/) to handle data propagation down the component stack and [re-rendering](https://stackoverflow.com/questions/69988412/how-can-i-re-render-my-component-when-a-property-to-it-changes-webcomponents) regardless of the order of component definition.

The preferred approach would be (2) since the functionality of our elements shouldn't be impacted by element definition order (easy to miss), however I'd like to implement (2) properly instead of building a hamstrung, one-off solution for the `table-stacked` property which will take a bit of time. In the meantime, I'm going to implement approach (1) to unblock the use of the responsive table.

See the 'reactive properties' task in #127 for implementing approach (2) above.

### Resources

* https://web.dev/articles/custom-elements-v1#element_upgrades
* https://html.spec.whatwg.org/multipage/custom-elements.html#custom-element-conformance
* https://html.spec.whatwg.org/multipage/custom-elements.html#custom-element-reactions
* https://javascript.info/slots-composition

**Comment by maxatdetroit** [comment]
A brief summary of the order of events in for a single web components:
1. Constructor
2. Attribute observer
3. Connected callback
4. Slot change events

A brief summary of the order of events for nested, slotted components:
```html
<parent my-attr="">
  <child>
    <grand-child>
    </grand-child>
  </child>
</parent>
```

Given all the context in this issue, the only safe way to propagate attribute `my-attr` all the way to the `grand-child` and actually react to its propagation, each component must:
1. Set the attribute on child components in the slot change event. E.g. parent sets `my-attr` on `child` during `slotchange` and `child` sets `my-attr` on `grand-child` during `slotchange`.
2. Observe the attribute change on self then do "some work" including setting the attribute on the child component - when `child` observes `my-attr` is sets `my-class` and then sets `my-attr` on `grand-child`

This may seem redundant, however, together they guard against various ordering of the element constructors being called which can happen if they are imported in an order different than their nesting and after the DOM is parsed. 

(1) handles the case when they are imported in the same order as their nesting. In that case, the slot change event should effectively propagate the attribute to child components.
(2) handles the case when they are imported in a different order than their nesting. In that case, the observe attribute handler continues the chain of propagation instead of the slotchange event.

**Comment by maxatdetroit** [comment]
This was resolved in #140.
